docs/style/CODING_STYLE_GUIDE.md

```diff
@@ -0,0 +1,172 @@
+
+## Language standard
+
+Code in the SDK conforms to the following standards. Changes to the C++ standard
+happen relatively infrequently. Changes to the Python version are more frequent.
+
+| Language | Version |
+| -------- | ------- |
+| C++      | C++17   |
+| Python   | 3.10    |
+
+Product-specific software may elect to use later standards in their own work.
+
+## Coding Guidelines
+
+### Common
+
+#### When in Rome
+
+The most important convention and practice in the Matter SDK repo is "_When in
+Rome..._", per the quote below.
+
+[quote, St. Ambrose]
+
+---
+
+If you should be in Rome, live in the Roman manner; if you should be elsewhere,
+live as they do there.
+
+---
+
+Your extensions or fixes to existing code should match the prevailing style of
+the original code.
+
+If you find the conventions so foreign or otherwise confusing, it may be best to
+let whoever owns the file make the necessary changes or seek the counsel of
+others in the group to find out what the right thing to do is. Never just start
+changing code wholesale for personal reasons without consulting others first.
+
+#### Commenting Out or Disabling Code
+
+Unused code shall not be disabled by commenting it out with C- or C++-style
+comments or with preprocessor `#if 0 ... #endif` semantics. Unused code should
+be removed.
+
+#### Auto-formatters
+
+We use the following auto-formatters on code:
+
+| Language    | Formatter          | Style File                                                                                 |
+| ----------- | ------------------ | ------------------------------------------------------------------------------------------ |
+| C++         | clang-format       | [.clang-format](https://github.com/project-chip/connectedhomeip/blob/master/.clang-format) |
+| Objective-C | clang-format       | [.clang-format](https://github.com/project-chip/connectedhomeip/blob/master/.clang-format) |
+| java        | google-java-format | N/A                                                                                        |
+| Python      | pep8, isort, ruff  | [.restyled.yaml][restyle_link] (command line), [isort][isort_link], [ruff][ruff_link]      |
+| YAML        | prettier           | None                                                                                       |
+| JSON        | prettier           | None                                                                                       |
+| markdown    | prettier           | None                                                                                       |
+
+[restyle_link]:
+    https://github.com/project-chip/connectedhomeip/blob/master/.restyled.yaml
+[isort_link]:
+    https://github.com/project-chip/connectedhomeip/blob/master/.isort.cfg
+[ruff_link]:
+    https://github.com/project-chip/connectedhomeip/blob/master/ruff.toml
+
+All pull requests run formatting checks using these tools before merge is
+allowed. Generated code is not run through restyle.
+
+### C++
```

Comment:
should we add a couple of sections on casting? something like:
- Use C++-style casts (e.g. static_cast) for type conversions instead of C-style casts.
- Whenever possible, use SafePointerCast.h instead of reinterpret_cast for pointer casting, since it is safer and adds alignment and type safety checks. 
- Avoid casting function pointers to `void *`, as this is considered undefined behavior in C++.

Reply:
Tracking new content suggestions in an issue for later: https://github.com/project-chip/connectedhomeip/issues/36544